### PR TITLE
ChatPromptTemplate: Expand support for message formats and documentation

### PR DESCRIPTION
**Pull request by eyurtsev**
* Expands support for a variety of message formats in the `from_messages` classmethod. Ideally, we could deprecate the other on-ramps to reduce the amount of classmethods users need to know about.
* Expand documentation with code examples.

### CODE REVIEW

File: libs/langchain/langchain/prompts/chat.py [hunk 1]
```diff
@@ -360,6 +359,67 @@ def from_strings(
         ]
         return cls.from_messages(messages)
 
+    @classmethod
+    def from_diverse(
```
Comment (eyurtsev) [inline libs/langchain/langchain/prompts/chat.py:363]: maybe someone has a better name (`from_generic`?)

**Comment by eyurtsev** [comment]
@hinthornw / @baskaryan should be ready to review. One question is whether we're okay mapping `assistant` to AIMessage / AIMessageTemplate if using a message representation of a 2-tuple `('assistant', 'Hello!')` equivalent to AIMessageTemplate(prompt='Hello!')

File: libs/langchain/langchain/prompts/chat.py [hunk 2]
```diff
@@ -261,63 +261,31 @@ def format_messages(self, **kwargs: Any) -> List[BaseMessage]:
...
-                    AIMessage(content="Hello!"), # Already formatted message
-                    HumanMessagePromptTemplate.from_template(
-                        "{user_input}"
-                    ),
-                ]
-            )
+            template = ChatPromptTemplate.from_messages([
+                ("system", "You are a helpful AI bot. Your name is {name}."),
+                ("human", "Hello, how are you doing?"),
+                ("assistant", "I'm doing well, thanks!"),
```
Comment (hwchase17) [inline libs/langchain/langchain/prompts/chat.py:277]: this should be "ai" now?

**Review by hwchase17 — APPROVED** [review]

File: libs/langchain/langchain/prompts/chat.py (hunk 2 above)
Comment (eyurtsev) [reply libs/langchain/langchain/prompts/chat.py:277]: yes! fixing in a follow up pr: https://github.com/langchain-ai/langchain/pull/8424/files

**Pull request merged**